features/verifysession/impl/src/main/kotlin/io/element/android/features/verifysession/impl/incoming/IncomingVerificationStateMachine.kt

```diff
@@ -0,0 +1,167 @@
+) : FlowReduxStateMachine<IncomingVerificationStateMachine.State, IncomingVerificationStateMachine.Event>(
+    initialState = State.Initial(isCancelled = false)
+) {
+    init {
+        spec {
+            inState<State.Initial> {
+                on { _: Event.AcceptIncomingRequest, state ->
+                    state.override { State.AcceptingIncomingVerification.andLog() }
+                }
+            }
+            inState<State.AcceptingIncomingVerification> {
+                onEnterEffect {
+                    sessionVerificationService.acceptVerificationRequest()
+                }
+                on { event: Event.DidReceiveChallenge, state ->
+                    state.override { State.ChallengeReceived(event.data).andLog() }
+                }
+            }
+            inState<State.ChallengeReceived> {
+                on { _: Event.AcceptChallenge, state ->
+                    state.override { State.AcceptingChallenge(state.snapshot.data).andLog() }
+                }
+                on { _: Event.DeclineChallenge, state ->
+                    state.override { State.RejectingChallenge(state.snapshot.data).andLog() }
+                }
+            }
+            inState<State.AcceptingChallenge> {
+                onEnterEffect { _ ->
+                    sessionVerificationService.approveVerification()
+                }
+                on { _: Event.DidAcceptChallenge, state ->
+                    state.override { State.Completed.andLog() }
+                }
+            }
+            inState<State.RejectingChallenge> {
+                onEnterEffect { _ ->
+                    sessionVerificationService.declineVerification()
+                }
+            }
+            inState<State.Canceling> {
+                onEnterEffect {
+                    sessionVerificationService.cancelVerification()
+                }
+            }
+            inState {
+                on { event: Event, state: MachineState<State> ->
+                    Timber.w("Verification in state [${state.snapshot::class.simpleName}] receiving event [${event::class.simpleName}]")
+                    state.noChange()
+                }
+                on { _: Event.Cancel, state: MachineState<State> ->
+                    when (state.snapshot) {
+                        State.Completed, State.Canceled -> state.noChange()
+                        else -> {
+                            sessionVerificationService.cancelVerification()
+                            state.override { State.Canceled.andLog() }
+                        }
+                    }
+                }
+                on { _: Event.DidCancel, state: MachineState<State> ->
+                    when (state.snapshot) {
+                        is State.RejectingChallenge -> {
+                            // For some reason, wait a bit, else the state machine is not happy...
+                            delay(100)
```

Comment:
Working to remove this line, but cannot figure out what's happening.
I I remove this patch and do a self session verification, then when the emoji are displayed click on "They do not match", I get a state machine error:

```
elementx: Verification in state [AcceptingIncomingVerification] receiving event [DidReceiveChallenge] | IncomingVerificationStateMachine.kt:69
elementx: Verification: moving state machine to new state [ChallengeReceived] | IncomingVerificationStateMachine.kt:107
elementx: Verification user action: DeclineVerification | IncomingVerificationPresenter.kt:86
elementx: Verification in state [ChallengeReceived] receiving event [DeclineChallenge] | IncomingVerificationStateMachine.kt:69
elementx: Verification: moving state machine to new state [RejectingChallenge] | IncomingVerificationStateMachine.kt:107
matrix_sdk::http_client::native: Sending request num_attempt=1 | crates/matrix-sdk/src/http_client/native.rs:55 | spans: send{server_versions=[V1_0, V1_1, V1_2, V1_3, V1_4, V1_5, V1_6, V1_7, V1_8, V1_9, V1_10, V1_11] config=RequestConfig { timeout: 30s } request_id="REQ-66" method=PUT uri="https://matrix-client.matrix.org/_matrix/client/v3/sendToDevice/m.key.verification.cancel/ebbca7912b0e4d43953960f2a4d5bbd3" request_size="165 B"}
elementx: Verification flow state: Canceled | IncomingVerificationPresenter.kt:162
elementx: Verification in state [RejectingChallenge] receiving event [DidCancel] | IncomingVerificationStateMachine.kt:69
elementx: Verification: moving state machine to new state [Failure] | IncomingVerificationStateMachine.kt:107
elementx: Failed to verify session
com.freeletics.flowredux.sideeffects.StateChangeCancellationException: StateMachine moved to a different state
	at com.freeletics.flowredux.sideeffects.ManagedSideEffect.cancelIfNeeded(SideEffect.kt:115)
	at com.freeletics.flowredux.FlowReduxKt$reduxStore$1$2$1.emit(FlowRedux.kt:42)
	at com.freeletics.flowredux.FlowReduxKt$reduxStore$1$2$1.emit(FlowRedux.kt:35)
	at kotlinx.coroutines.flow.FlowKt__ChannelsKt.emitAllImpl$FlowKt__ChannelsKt(Channels.kt:33)
	at kotlinx.coroutines.flow.FlowKt__ChannelsKt.access$emitAllImpl$FlowKt__ChannelsKt(Channels.kt:1)
	at kotlinx.coroutines.flow.FlowKt__ChannelsKt$emitAllImpl$1.invokeSuspend(Unknown Source:14)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:101)
	at androidx.compose.ui.platform.AndroidUiDispatcher.performTrampolineDispatch(AndroidUiDispatcher.android.kt:81)
	at androidx.compose.ui.platform.AndroidUiDispatcher.access$performTrampolineDispatch(AndroidUiDispatcher.android.kt:41)
	at androidx.compose.ui.platform.AndroidUiDispatcher$dispatchCallback$1.run(AndroidUiDispatcher.android.kt:57)
	at android.os.Handler.handleCallback(Handler.java:958)
	at android.os.Handler.dispatchMessage(Handler.java:99)
	at android.os.Looper.loopOnce(Looper.java:205)
	at android.os.Looper.loop(Looper.java:294)
	at android.app.ActivityThread.main(ActivityThread.java:8177)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:552)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:971)
 | RustSessionVerificationService.kt:150
elementx: Session verification failed with an unknown error | RustSessionVerificationService.kt:167
elementx: Verification flow state: Failed | IncomingVerificationPresenter.kt:162
elementx: Verification in state [Failure] receiving event [DidFail] | IncomingVerificationStateMachine.kt:69
elementx: Verification: moving state machine to new state [Failure] | IncomingVerificationStateMachine.kt:107
matrix_sdk_crypto::machine: Received a to-device event | crates/matrix-sdk-crypto/src/machine/mod.rs:1277 | spans: next_sync_with_lock > sync_once{pos="0/s5351332706_757284974_8456330_3193155361_3469265458_261604925_1398669885_11029814222_0_359034"} > handle_response > preprocess_to_device_events > receive_sync_changes > receive_to_device_event{sender="@benoitx:matrix.org" event_type="m.key.verification.cancel" }
```

Reply:
Fixed in [63b31ac](https://github.com/element-hq/element-x-android/pull/3733/commits/63b31ac8b174ffff7b73ef1519bb8c796e92925a)